File: CS/WindowsUIDemo.Wpf/Modules/HamburgerMenuDemo/CompactModeConverters.cs
using System;
using System.Globalization;
using System.IO;
using System.Windows.Data;
using System.Windows.Documents;
using System.Windows.Media;
using System.Windows.Media.Imaging;
using DevExpress.Data;

namespace WindowsUIDemo {
    public class OrderNameConverter : IMultiValueConverter {
        public object Convert(object[] values, Type targetType, object parameter, CultureInfo culture) {
            if(values.Length != 2)
                return null;
            if(!(values[1] is ColumnSortOrder))
                return null;
            ColumnSortOrder order = (ColumnSortOrder)values[1];
            if(values[0].ToString() == "Received") {
                switch(order) {
                    case ColumnSortOrder.Ascending:
                        return "Oldest";
                    case ColumnSortOrder.Descending:
                        return "Newest";
                    default:
                        return null;
                }
            }
            switch(order) {
                case ColumnSortOrder.Ascending:
                    return "Ascending";
                case ColumnSortOrder.Descending:
                    return "Descending";
                default:
                    return null;
            }
        }

        public object[] ConvertBack(object value, Type[] targetTypes, object parameter, CultureInfo culture) {
            throw new NotImplementedException();
        }
    }

    public class MessageToFillConverter : IValueConverter {
        public object Convert(object value, Type targetType, object parameter, CultureInfo culture) {
            var message = value as Message;
            if(message == null) return null;
            if (message.Employee.Image == null) {
                string name = message.From;
                if(string.IsNullOrEmpty(name))
                    return Colors.Black;
                byte color = (byte)(Math.Abs(name.GetHashCode()) % 5);
                switch(color) {
                    case 0:
                        return new SolidColorBrush((Color)ColorConverter.ConvertFromString("#FF31669C"));
                    case 1:
                        return new SolidColorBrush((Color)ColorConverter.ConvertFromString("#FF598A7A"));
                    case 2:
                        return new SolidColorBrush((Color)ColorConverter.ConvertFromString("#FFCCA464"));
                    case 3:
                        return new SolidColorBrush((Color)ColorConverter.ConvertFromString("#FFE06B4C"));
                    case 4:
                        return new SolidColorBrush((Color)ColorConverter.ConvertFromString("#FF558AC0"));
                    default:
                        return new SolidColorBrush(Colors.Black);
                }
            } else {
                byte[] imageData = message.Employee.Image as byte[];
                if(imageData == null)
                    return null;
                BitmapImage biImg = new BitmapImage();
                MemoryStream ms = new MemoryStream(imageData);
                biImg.BeginInit();
                biImg.StreamSource = ms;
                biImg.EndInit();
                ImageSource imgSrc = biImg as ImageSource;

                return new ImageBrush(imgSrc);
            }
        }

        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture) {
            throw new NotImplementedException();
        }
    }

    public class NameToInitialsConverter : IValueConverter {
        public object Convert(object value, Type targetType, object parameter, CultureInfo culture) {
            string name = value as string;
            if(string.IsNullOrEmpty(name))
                return "JJ";
            int idx = name.IndexOf(' ');
            if(idx < 0 || name.Length <= idx + 1 || name[idx + 1] == '(')
                return name[0].ToString().ToUpper();
            return (name[0].ToString() + name[idx + 1].ToString()).ToUpper();
        }

        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture) {
            throw new NotImplementedException();
        }
    }

    public class PriorityGroupValueConverter : IValueConverter {
        public object Convert(object value, Type targetType, object parameter, CultureInfo culture) {
            if(value is int) {
                int priority = (int)value;
                switch(priority) {
                    case 0:
                        return "Low priority";
                    case 1:
                        return "Normal priority";
                    case 2:
                        return "Hight priority";
                    default:
                        return null;
                }
            }
            return null;
        }

        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture) {
            throw new NotImplementedException();
        }
    }

    public class AttachmentGroupValueConverter : IValueConverter {
        public object Convert(object value, Type targetType, object parameter, CultureInfo culture) {
            if(value is bool) {
                bool attachment = (bool)value;
                return attachment ? "Has Attachment" : "Hasn't Attachment";
            }
            return null;
        }

        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture) {
            throw new NotImplementedException();
        }
    }
    public class ByteImageConverter : IValueConverter {

        public int DecodePixelHeight { get; set; }

        public object Convert(object value, Type targetType, object parameter, CultureInfo culture) {
            byte[] imageData = value as byte[];
            if(imageData == null)
                return null;
            BitmapImage biImg = new BitmapImage();
            RenderOptions.SetBitmapScalingMode(biImg, BitmapScalingMode.HighQuality);
            MemoryStream ms = new MemoryStream(imageData);
            biImg.BeginInit();
            biImg.DecodePixelHeight = DecodePixelHeight;
            biImg.StreamSource = ms;
            biImg.EndInit();
            ImageSource imgSrc = biImg as ImageSource;
            return imgSrc;
        }
        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture) {
            throw new NotImplementedException();
        }
    }
    public class NameToColorConverter : IValueConverter {
        public object Convert(object value, Type targetType, object parameter, CultureInfo culture) {
            string name = value as string;
            if(string.IsNullOrEmpty(name))
                return Colors.Black;
            byte color = (byte)(Math.Abs(name.GetHashCode()) % 5);
            switch(color) {
                case 0:
                    return (Color)ColorConverter.ConvertFromString("#FF31669C");
                case 1:
                    return (Color)ColorConverter.ConvertFromString("#FF598A7A");
                case 2:
                    return (Color)ColorConverter.ConvertFromString("#FFCCA464");
                case 3:
                    return (Color)ColorConverter.ConvertFromString("#FFE06B4C");
                case 4:
                    return (Color)ColorConverter.ConvertFromString("#FF558AC0");
                default:
                    return Colors.Black;
            }
        }

        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture) {
            throw new NotImplementedException();
        }
    }

}

File: VB/WindowsUIDemo.Wpf/Modules/HamburgerMenuDemo/CompactModeConverters.vb
Imports System
Imports System.Globalization
Imports System.IO
Imports System.Windows.Data
Imports System.Windows.Documents
Imports System.Windows.Media
Imports System.Windows.Media.Imaging
Imports DevExpress.Data

Namespace WindowsUIDemo
    Public Class OrderNameConverter
        Implements IMultiValueConverter

        Public Function Convert(ByVal values() As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IMultiValueConverter.Convert
            If values.Length <> 2 Then
                Return Nothing
            End If
            If Not(TypeOf values(1) Is ColumnSortOrder) Then
                Return Nothing
            End If
            Dim order As ColumnSortOrder = DirectCast(values(1), ColumnSortOrder)
            If values(0).ToString() = "Received" Then
                Select Case order
                    Case ColumnSortOrder.Ascending
                        Return "Oldest"
                    Case ColumnSortOrder.Descending
                        Return "Newest"
                    Case Else
                        Return Nothing
                End Select
            End If
            Select Case order
                Case ColumnSortOrder.Ascending
                    Return "Ascending"
                Case ColumnSortOrder.Descending
                    Return "Descending"
                Case Else
                    Return Nothing
            End Select
        End Function

        Public Function ConvertBack(ByVal value As Object, ByVal targetTypes() As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object() Implements IMultiValueConverter.ConvertBack
            Throw New NotImplementedException()
        End Function
    End Class

    Public Class MessageToFillConverter
        Implements IValueConverter

        Public Function Convert(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.Convert
            Dim message = TryCast(value, Message)
            If message Is Nothing Then
                Return Nothing
            End If
            If message.Employee.Image Is Nothing Then
                Dim name As String = message.From
                If String.IsNullOrEmpty(name) Then
                    Return Colors.Black
                End If
                Dim color As Byte = CByte(Math.Abs(name.GetHashCode()) Mod 5)
                Select Case color
                    Case 0
                        Return New SolidColorBrush(DirectCast(ColorConverter.ConvertFromString("#FF31669C"), Color))
                    Case 1
                        Return New SolidColorBrush(DirectCast(ColorConverter.ConvertFromString("#FF598A7A"), Color))
                    Case 2
                        Return New SolidColorBrush(DirectCast(ColorConverter.ConvertFromString("#FFCCA464"), Color))
                    Case 3
                        Return New SolidColorBrush(DirectCast(ColorConverter.ConvertFromString("#FFE06B4C"), Color))
                    Case 4
                        Return New SolidColorBrush(DirectCast(ColorConverter.ConvertFromString("#FF558AC0"), Color))
                    Case Else
                        Return New SolidColorBrush(Colors.Black)
                End Select
            Else
                Dim imageData() As Byte = TryCast(message.Employee.Image, Byte())
                If imageData Is Nothing Then
                    Return Nothing
                End If
                Dim biImg As New BitmapImage()
                Dim ms As New MemoryStream(imageData)
                biImg.BeginInit()
                biImg.StreamSource = ms
                biImg.EndInit()
                Dim imgSrc As ImageSource = TryCast(biImg, ImageSource)

                Return New ImageBrush(imgSrc)
            End If
        End Function

        Public Function ConvertBack(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
            Throw New NotImplementedException()
        End Function
    End Class

    Public Class NameToInitialsConverter
        Implements IValueConverter

        Public Function Convert(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.Convert
            Dim name As String = TryCast(value, String)
            If String.IsNullOrEmpty(name) Then
                Return "JJ"
            End If
            Dim idx As Integer = name.IndexOf(" "c)
            If idx < 0 OrElse name.Length <= idx + 1 OrElse name.Chars(idx + 1) = "("c Then
                Return name.Chars(0).ToString().ToUpper()
            End If
            Return (name.Chars(0).ToString() & name.Chars(idx + 1).ToString()).ToUpper()
        End Function

        Public Function ConvertBack(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
            Throw New NotImplementedException()
        End Function
    End Class

    Public Class PriorityGroupValueConverter
        Implements IValueConverter

        Public Function Convert(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.Convert
            If TypeOf value Is Integer Then
                Dim priority As Integer = DirectCast(value, Integer)
                Select Case priority
                    Case 0
                        Return "Low priority"
                    Case 1
                        Return "Normal priority"
                    Case 2
                        Return "Hight priority"
                    Case Else
                        Return Nothing
                End Select
            End If
            Return Nothing
        End Function

        Public Function ConvertBack(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
            Throw New NotImplementedException()
        End Function
    End Class

    Public Class AttachmentGroupValueConverter
        Implements IValueConverter

        Public Function Convert(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.Convert
            If TypeOf value Is Boolean Then
                Dim attachment As Boolean = DirectCast(value, Boolean)
                Return If(attachment, "Has Attachment", "Hasn't Attachment")
            End If
            Return Nothing
        End Function

        Public Function ConvertBack(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
            Throw New NotImplementedException()
        End Function
    End Class
    Public Class ByteImageConverter
        Implements IValueConverter

        Public Property DecodePixelHeight() As Integer

        Public Function Convert(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.Convert
            Dim imageData() As Byte = TryCast(value, Byte())
            If imageData Is Nothing Then
                Return Nothing
            End If
            Dim biImg As New BitmapImage()
            RenderOptions.SetBitmapScalingMode(biImg, BitmapScalingMode.HighQuality)
            Dim ms As New MemoryStream(imageData)
            biImg.BeginInit()
            biImg.DecodePixelHeight = DecodePixelHeight
            biImg.StreamSource = ms
            biImg.EndInit()
            Dim imgSrc As ImageSource = TryCast(biImg, ImageSource)
            Return imgSrc
        End Function
        Public Function ConvertBack(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
            Throw New NotImplementedException()
        End Function
    End Class
    Public Class NameToColorConverter
        Implements IValueConverter

        Public Function Convert(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.Convert
            Dim name As String = TryCast(value, String)
            If String.IsNullOrEmpty(name) Then
                Return Colors.Black
            End If
            Dim color As Byte = CByte(Math.Abs(name.GetHashCode()) Mod 5)
            Select Case color
                Case 0
                    Return DirectCast(ColorConverter.ConvertFromString("#FF31669C"), Color)
                Case 1
                    Return DirectCast(ColorConverter.ConvertFromString("#FF598A7A"), Color)
                Case 2
                    Return DirectCast(ColorConverter.ConvertFromString("#FFCCA464"), Color)
                Case 3
                    Return DirectCast(ColorConverter.ConvertFromString("#FFE06B4C"), Color)
                Case 4
                    Return DirectCast(ColorConverter.ConvertFromString("#FF558AC0"), Color)
                Case Else
                    Return Colors.Black
            End Select
        End Function

        Public Function ConvertBack(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
            Throw New NotImplementedException()
        End Function
    End Class

End Namespace
